velox/exec/RowsStreamingWindowBuild.cpp

```diff
@@ -101,13 +109,29 @@ void RowsStreamingWindowBuild::noMoreInput() {
 }
 
 std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::nextPartition() {
+  // Remove the processed output partition from the queue.
+  //
+  // NOTE: the window operator only calls this after processing a completed
+  // partition.
+  if (!windowPartitions_.empty() && windowPartitions_.front()->complete() &&
+      windowPartitions_.front()->numRows() == 0) {
+    windowPartitions_.pop_front();
+  }
+
   VELOX_CHECK(hasNextPartition());
-  return windowPartitions_[++outputPartition_];
+  return windowPartitions_.front();
 }
 
 bool RowsStreamingWindowBuild::hasNextPartition() {
-  return !windowPartitions_.empty() &&
-      outputPartition_ + 2 <= windowPartitions_.size();
+  // Checks if there is a window partition that is either incomplete or
+  // completed but has unconsumed rows.
+  for (const auto& windowPartition : windowPartitions_) {
```

Comment:
nit: Probably not a big deal, but it looks like the current code iterates through windowPartitions_ from the beginning to the end, while new partitions that are likely incomplete are added from the back, so it may make more sense to loop from the back to the beginning instead? (Please correct me if I'm wrong. Thanks!)

Reply:
@kagamiori Good catch. It makes sense to me. I have updated the code. Thanks for your review .